src/clients/base.rs

```diff
@@ -0,0 +1,1054 @@
+            })
+            .filter_map(
+                // TODO: not sure if this `to_string` is what we want. It
+                // might add quotes to the strings.
+                |param| payload.get(&param).map(|value| (param, value.to_string())),
+            )
+            .collect::<HashMap<_, _>>();
+
+        for (ref key, ref value) in &map_to_hold_owned_value {
+            params.insert(key, value);
+        }
+
+        let result = self.endpoint_get("recommendations", &params).await?;
+        convert_result(&result)
+    }
+
+    /// Get full details of the tracks of a playlist owned by a user.
+    ///
+    /// Parameters:
+    /// - playlist_id - the id of the playlist
+    /// - fields - which fields to return
+    /// - limit - the maximum number of tracks to return
+    /// - offset - the index of the first track to return
+    /// - market - an ISO 3166-1 alpha-2 country code or the string from_token.
+    ///
+    /// See [`Self::playlist_tracks`] for a manually paginated version of this.
+    ///
+    /// [Reference](https://developer.spotify.com/documentation/web-api/reference/#endpoint-get-playlists-tracks)
+    fn playlist_tracks<'a>(
+        &'a self,
+        playlist_id: &'a PlaylistId,
+        fields: Option<&'a str>,
+        market: Option<&'a Market>,
+    ) -> Paginator<'_, ClientResult<PlaylistItem>> {
+        paginate(
+            move |limit, offset| {
+                self.playlist_tracks_manual(playlist_id, fields, market, Some(limit), Some(offset))
+            },
+            self.get_config().pagination_chunks,
+        )
+    }
+
+    /// The manually paginated version of [`Self::playlist_tracks`].
+    async fn playlist_tracks_manual(
+        &self,
+        playlist_id: &PlaylistId,
+        fields: Option<&str>,
+        market: Option<&Market>,
+        limit: Option<u32>,
+        offset: Option<u32>,
+    ) -> ClientResult<Page<PlaylistItem>> {
+        let limit = limit.map(|s| s.to_string());
+        let offset = offset.map(|s| s.to_string());
+        let params = build_map! {
+            optional "fields": fields,
+            optional "market": market.map(|x| x.as_ref()),
+            optional "limit": limit.as_deref(),
+            optional "offset": offset.as_deref(),
+        };
+
+        let url = format!("playlists/{}/tracks", playlist_id.id());
+        let result = self.endpoint_get(&url, &params).await?;
+        convert_result(&result)
+    }
+
+    /// Gets playlists of a user.
+    ///
+    /// Parameters:
+    /// - user_id - the id of the usr
+    /// - limit  - the number of items to return
+    /// - offset - the index of the first item to return
+    ///
+    /// See [`Self::user_playlists_manual`] for a manually paginated version of
+    /// this.
+    ///
+    /// [Reference](https://developer.spotify.com/documentation/web-api/reference/#endpoint-get-list-users-playlists)
+    fn user_playlists<'a>(
```

Comment:
The `user_playlists` and `user_playlists_manual` endpoints should be put into the `oauth.rs` instead of base.rs. Since according to the document, 
private playlists are only retrievable for the current user and requires the `playlist-read-private` scope to have been authorized by the user

Reply:
The thing with these endpoints you've mentioned is that, although it may work with OAuth, it *can also be used without it*. So they must be accessible from the base trait. Check out its reference:

![image](https://user-images.githubusercontent.com/25647296/123631070-9e33b100-d816-11eb-855b-e392b1ec7e3d.png)

And the footnote:

![image](https://user-images.githubusercontent.com/25647296/123631499-18fccc00-d817-11eb-8a53-69bb02646407.png)

You can try it out here: https://developer.spotify.com/console/get-playlists/?user_id=wizzler&limit=&offset=

Reply:
Sorry for not checking the footnote and its reference carefully, great point!